synapse/lib/scope.py

```diff
@@ -179,5 +175,52 @@ def enter(vals=None):
     '''
     scope = _task_scope()
     scope.enter(vals)
-    yield
-    scope.leave()
+    try:
+        yield
+    finally:
+        scope.leave()
+
+def copy(task: asyncio.Task) -> None:
+    '''
+    Clone the current task Scope onto the provided task.
+
+    Args:
+        task (asyncio.Task): The task object to attach the scope too.
+
+    Notes:
+        This must be run from an asyncio IO loop.
+
+        If the current task does not have a scope, we clone the default global Scope.
+
+        This will ``enter()`` the scope, and add a task callback to ``leave()`` the scope.
+
+    Returns:
+        None
+    '''
+
+    current_task = asyncio.current_task()
+
+    if current_task is None:
+        # It is possible that we are executing code started by
+        # asyncio.call_soon_threadsafe (or similar mechanisms)
+        # in which case there is not yet a task for us to
+        # retrieve the scope from, and we can inherit directly
+        # from globscope.
+
+        parent_scope = globscope
+
+    else:
+
+        parent_scope = _task_scope()
+
+    scope = parent_scope.copy()
+    task._syn_scope = scope
+    scope.enter()
+
+    def fini(_task):
```

Comment:
lets name this callback `done` or `taskdone` to avoid confusion with our `fini()` functions